storm-client/src/jvm/org/apache/storm/daemon/Task.java

```diff
@@ -122,28 +130,33 @@ public Task(Executor executor, Integer taskId) throws IOException {
         return new ArrayList<>(0);
     }
 
+
     public List<Integer> getOutgoingTasks(String stream, List<Object> values) {
         if (debug) {
             LOG.info("Emitting Tuple: taskId={} componentId={} stream={} values={}", taskId, componentId, stream, values);
         }
 
-        List<Integer> outTasks = new ArrayList<>();
-        if (!streamComponentToGrouper.containsKey(stream)) {
-            throw new IllegalArgumentException("Unknown stream ID: " + stream);
-        }
-        if (null != streamComponentToGrouper.get(stream)) {
-            // null value for __system
-            for (LoadAwareCustomStreamGrouping grouper : streamComponentToGrouper.get(stream).values()) {
+        ArrayList<Integer> outTasks = new ArrayList<>();
+
+        // TODO: PERF: expensive hashtable lookup in critical path
```

Comment:
Yes hashmap lookups showed up as an important throughput bottleneck during profiling. Had to spend a bunch of time eliminating hashmap usage in critical path. a few still remain unfortunately.